### PR TITLE
Hive engine compatible with Hive 2.3

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Try to make the Hive engine compatible with Hive 2.3.

It is not easy to align all dependencies with Hive 2.3, thus this PR just added a new integration test to make sure that `kyuubi hive engine` compiled against Hive 3.1.3 could work with Hive 2.3.9 runtime, instead of introducing new a profile to make it pass building with Hive 2.3.9

There are potential class conflict issues, because the Apache Hive 2.3.9 and 3.1.3 official binary artifacts ship Scala 2.11 jars but Kyuubi uses Scala 2.12.

Also tested with `apache-hive-2.1.1-cdh6.3.2-bin.tar.gz` locally, which can successfully bootstrap and run queries.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/hive/HiveProcessBuilder.scala [hunk 1]
```diff
@@ -116,8 +116,6 @@ class HiveProcessBuilder(
     buffer.toArray
   }
 
-  override def toString: String = Utils.redactCommandLineArgs(conf, commands).mkString("\n")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/hive/HiveProcessBuilder.scala:119]: the output is ugly, prefer to use the parent implementation

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -261,8 +266,15 @@ jobs:
       - name: Build and test Hive with maven w/o linters
         run: |
           TEST_MODULES="externals/kyuubi-hive-sql-engine,integration-tests/kyuubi-hive-it"
-          ./build/mvn ${MVN_OPT} -pl ${TEST_MODULES} -am clean install -DskipTests
-          ./build/mvn ${MVN_OPT} -pl ${TEST_MODULES} test
+          ./build/mvn ${MVN_OPT} ${{ matrix.hive-archive }} -pl ${TEST_MODULES} -am clean install -DskipTests
+          # Hive 2.3.9 ships Derby 10.10.2.0, which may fail to boostrap on latest JDK 8
+          # https://github.com/apache/hive/pull/4895
+          if [[ "${{ matrix.hive-archive }}" == *apache-hive-2.3.9-bin.tar.gz* ]]; then
+            HIVE_239_LIB="$PWD/externals/kyuubi-download/target/apache-hive-2.3.9-bin/lib"
```
Comment (Yikf) [inline .github/workflows/master.yml:273]: Nice job! I suggest that to record this information in the official document.

Comment (pan3793) [reply .github/workflows/master.yml:273]: This only affects users who use derby as the embedded metastore storage backend, not a production solution

Comment (pan3793) [reply .github/workflows/master.yml:273]: but it worth to write a dedicated docs page for hive engines deployment

Comment (Yikf) [reply .github/workflows/master.yml:273]: > This only affects users who use derby as the embedded metastore storage backend, not a production solution

Get

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master/1.8

**Pull request closed without merging**